src/sdk/trace-metrics.ts

```diff
@@ -153,36 +153,85 @@ export class ExperienceMetrics {
 }
 
 export class CumulativeLayoutShift {
-  static computeCLSValue(events: Array<TraceEvent>) {
-    const layoutShiftEvents = events.filter(
-      event => event.name === 'LayoutShift' && event.args?.data.is_main_frame
-    );
+  static getLayoutShiftEvents(traceEvents: Array<TraceEvent>) {
+    const layoutShiftEvents = [];
     // Chromium will set `had_recent_input` if there was recent user input, which
     // skips shift events from contributing to CLS. This results in the first few shift
     // events having `had_recent_input` set to true, so ignore it for those events.
-    // See https://bugs.chromium.org/p/chromium/issues/detail?id=1094974.
+    // See https://crbug.com/1094974
     let ignoreHadRecentInput = true;
-    let finalLayoutShiftEvent;
-    let clsScore = 0;
-    for (const event of layoutShiftEvents) {
+
+    for (const event of traceEvents) {
+      // weighted_score_delta was added in chrome 90 - https://crbug.com/1173139
+      if (
+        event.name !== 'LayoutShift' ||
+        !event.args?.data.is_main_frame ||
+        !event.args?.data.weighted_score_delta
+      ) {
+        continue;
+      }
+
       if (event.args.data.had_recent_input) {
         if (!ignoreHadRecentInput) continue;
       } else {
         ignoreHadRecentInput = false;
       }
-      clsScore += event.args.data.score;
-      finalLayoutShiftEvent = event;
+
+      layoutShiftEvents.push({
+        ts: event.ts,
+        weightedScore: event.args.data.weighted_score_delta,
+      });
     }
 
-    return { score: clsScore, event: finalLayoutShiftEvent };
+    return layoutShiftEvents;
+  }
+
+  /**
+   * Calculate cumulative layout shift per sesion window where each session
+   * windows lasts for 5 seconds since the last layoutShift event and has 1 second
+   * gap between them. Return the maximum score between all windows.
+   * More details - https://web.dev/evolving-cls/
+   */
+  static calculateScore(layoutShiftEvents) {
+    const gapMicroseconds = 1_000_000;
+    const limitMicroseconds = 5_000_000;
+    let maxScore = 0;
+    let currentScore = 0;
+    let firstTimestamp = Number.NEGATIVE_INFINITY;
+    let prevTimestamp = Number.NEGATIVE_INFINITY;
+
+    for (const event of layoutShiftEvents) {
+      if (
+        event.ts - firstTimestamp > limitMicroseconds ||
```

Comment:
The logic here is tricky, and comments always help I think. It's doing two things in one pass, calculating the current window, and figuring out which window is to be used.  The code would be clearer IMHO if this was broken down more. Maybe even using map/reduce (esp. since this isn't perf critical code). It might be clearer to have two distinct phases, creating the windows, then selecting one, rather than doing it all-in-one.

Reply:
Kept in sync with lighthouse so we can use keep it more or less same and get rid of our code at some point. I can add more comments if that helps.

Reply:
Added detailed comments, Should be good to reason about it now. I don't think we need to create session windows manually and use map-reduce. The current code feels cleaner to me and can accomplish it in a single pass.